akka-http-scala/src/main/scala/akka/http/scaladsl/server/directives/SecurityDirectives.scala

```diff
@@ -5,10 +5,111 @@
 package akka.http.scaladsl.server
 package directives
 
-import BasicDirectives._
-import RouteDirectives._
+import scala.reflect.ClassTag
+import scala.concurrent.Future
+import akka.http.impl.util._
+import akka.http.scaladsl.util.FastFuture
+import akka.http.scaladsl.util.FastFuture._
+import akka.http.scaladsl.model.headers._
+import akka.http.scaladsl.server.AuthenticationFailedRejection.{ CredentialsRejected, CredentialsMissing }
 
+/**
+ * Provides directives for securing an inner route using the standard Http authentication headers [[`WWW-Authenticate`]]
+ * and [[Authorization]]. Most prominently, HTTP Basic authentication as defined in RFC 2617.
+ */
 trait SecurityDirectives {
+  import BasicDirectives._
+  import HeaderDirectives._
+  import FutureDirectives._
+  import RouteDirectives._
+
+  /**
+   * The result of an HTTP authentication attempt is either the user object or
+   * an HttpChallenge to present to the browser.
+   */
+  type AuthenticationResult[+T] = Either[HttpChallenge, T]
+
+  type Authenticator[T] = UserCredentials ⇒ Future[Option[T]]
+  type StrictAuthenticator[T] = UserCredentials ⇒ Option[T]
+  type AuthenticatorPF[T] = PartialFunction[UserCredentials, Future[T]]
+  type StrictAuthenticatorPF[T] = PartialFunction[UserCredentials, T]
+
+  /**
+   * Extracts the potentially present [[HttpCredentials]] provided with the request's [[Authorization]] header.
+   */
+  def extractCredentials: Directive1[Option[HttpCredentials]] =
+    optionalHeaderValueByType[Authorization]().map(_.map(_.credentials))
+
+  /**
+   * A directive that wraps the inner route with Http Basic authentication support.
+   * The given authenticator determines whether the credentials in the request are valid
+   * and, if so, which user object to supply to the inner route.
+   */
+  def authenticateBasic[T](realm: String, authenticator: Authenticator[T]): AuthenticationDirective[T] =
+    extractExecutionContext.flatMap { implicit ec ⇒
+      authenticateOrRejectWithChallenge[BasicHttpCredentials, T] { basic ⇒
+        authenticator(UserCredentials(basic)).fast.map {
+          case Some(t) ⇒ AuthenticationResult.success(t)
+          case None    ⇒ AuthenticationResult.failWithChallenge(challengeFor(realm))
+        }
+      }
+    }
+
+  /**
+   * A directive that wraps the inner route with Http Basic authentication support.
+   * The given authenticator determines whether the credentials in the request are valid
+   * and, if so, which user object to supply to the inner route.
+   */
+  def authenticateBasicStrict[T](realm: String, authenticator: StrictAuthenticator[T]): AuthenticationDirective[T] =
```

Comment:
I'll switch the terminology around and make it `authenticateBasic` and name the other `authenticateBasicAsync` which also follows the akka-stream convention.